### DISCUSSION TITLE
ERROR: member access into incomplete type `llama_context`

### DISCUSSION THREAD

**Discussion by walker-ai** [ggml-org/llama.cpp#9404] [Q&A] [answered]
I added several member variable(most of them are struct pointer), and I tried to access or assign them in the `main.cpp`, and then I receive the error message like this:
![image](https://github.com/user-attachments/assets/46084a60-a11e-4d7c-bc65-ad742521a227)

I checked the original content of `main.cpp` and I found that there is no such thing as writing like I do (access or assign member of `llama_context`). Is the reason that it can't be written that way, or is it that I'm not writing it correctly somewhere? 😭🙏

**Comment by ngxson** [comment, marked as answer]
`llama_context` is an internal struct (non-public), you cannot access its members from outside of `llama.cpp`

**Reply by walker-ai** [reply]
I see that, thanks😊

**Reply by walker-ai** [reply]
Btw, If I want to pass some parameters to `llama.cpp`...

> Which are used when building the llama graph, and since the `llama_context* ctx` has been used throughout the build, I'd naturally think to add some custom parameters to it.

But these parameters are related to the user-entered prompt, so I'd only be able to access or assign them in main.cpp, could you give me some suggestions? 🙏

**Reply by ngxson** [reply]
One hacky way would be to define your own struct in `llama.cpp`

```cpp
struct my_own_data {
    ...
};
```

Then add it as member of `struct llama_context`, something like `ctx->my_data`

Finally, add a getter function for it in llama.cpp/.h

```cpp
struct my_own_data * get_my_own_data(llama_context * ctx) {
  return &ctx->my_data;
}
```

**Reply by ngxson** [reply]
The proper way is to make a wrapper outside llama.cpp, then bring your own functions there. It's wrapper design pattern

**Reply by walker-ai** [reply]
> One hacky way would be to define your own struct in `llama.cpp`
> 
> ```c++
> struct my_own_data {
>     ...
> };
> ```
> 
> Then add it as member of `struct llama_context`, something like `ctx->my_data`
> 
> Finally, add a getter function for it in llama.cpp/.h
> 
> ```c++
> struct my_own_data * get_my_own_data(llama_context * ctx) {
>   return &ctx->my_data;
> }
> ```

I will try that, thx a lot
